jni/src/faiss_index_service.cpp

```diff
@@ -137,27 +233,34 @@ void BinaryIndexService::createIndex(
         throw std::runtime_error("Number of IDs does not match number of vectors");
     }
 
-    std::unique_ptr<faiss::IndexBinary> indexWriter(faissMethods->indexBinaryFactory(dim, indexDescription.c_str()));
-
     // Set thread count if it is passed in as a parameter. Setting this variable will only impact the current thread
     if(threadCount != 0) {
         omp_set_num_threads(threadCount);
     }
 
-    // Add extra parameters that cant be configured with the index factory
-    SetExtraParameters<faiss::IndexBinary, faiss::IndexBinaryIVF, faiss::IndexBinaryHNSW>(jniUtil, env, parameters, indexWriter.get());
-
-    // Check that the index does not need to be trained
-    if(!indexWriter->is_trained) {
-        throw std::runtime_error("Index is not trained");
-    }
+    faiss::IndexBinaryIDMap * idMap = reinterpret_cast<faiss::IndexBinaryIDMap *> (idMapAddress);
 
     // Add vectors
-    std::unique_ptr<faiss::IndexBinaryIDMap> idMap(faissMethods->indexBinaryIdMap(indexWriter.get()));
     idMap->add_with_ids(numVectors, inputVectors->data(), ids.data());
+}
+
+void BinaryIndexService::writeIndex(
+        int threadCount,
+        std::string indexPath,
+        jlong idMapAddress
+    ) {
+    // Set thread count if it is passed in as a parameter. Setting this variable will only impact the current thread
+    if(threadCount != 0) {
+        omp_set_num_threads(threadCount);
+    }
+
+    std::unique_ptr<faiss::IndexBinaryIDMap> idMap (reinterpret_cast<faiss::IndexBinaryIDMap *> (idMapAddress));
 
     // Write the index to disk
     faissMethods->writeIndexBinary(idMap.get(), indexPath.c_str());
+
+    // Free the memory used by the index
+    delete idMap->index;
```

Comment:
same here, we should make sure we free up the memory when there is an exception

Reply:
How should I free the memory if delete is not working?

Reply:
wrap the call with try catch and delete it in catch

https://github.com/opensearch-project/k-NN/blob/main/jni/src/nmslib_wrapper.cpp#L269

Reply:
Ah. Are you saying to wrap the whole function in try catch?

Reply:
Added. Let me know if I did it right. Resolving for now.